File: backend/app/models/service.py
import uuid
from typing import Optional

from sqlalchemy import Float, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, UUIDMixin


class Service(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "services"

    port_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("ports.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[Optional[str]] = mapped_column(String(255), nullable=True, index=True)
    product: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    version: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    extra_info: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    tunnel: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    protocol: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    banner: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    normalized_name: Mapped[Optional[str]] = mapped_column(String(255), nullable=True, index=True)
    normalized_product: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    normalized_version: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    category: Mapped[Optional[str]] = mapped_column(String(100), nullable=True, index=True)
    confidence: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    port: Mapped["Port"] = relationship("Port", back_populates="services")

File: frontend/src/types/service.ts
export interface ServiceIntelligence {
  id: string;
  port_id: string;
  name: string | null;
  product: string | null;
  version: string | null;
  extra_info: string | null;
  tunnel: string | null;
  protocol: string | null;
  banner: string | null;
  normalized_name: string | null;
  normalized_product: string | null;
  normalized_version: string | null;
  category: string | null;
  confidence: number | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
  port_number: number | null;
  port_protocol: string | null;
  host_id: string | null;
  host_ip: string | null;
  host_name: string | null;
}

export interface ServiceEnrichRequest {
  service_ids?: string[];
}
